File: NET/Libraries/Treehopper.Libraries/Displays/Ht16k33.cs
namespace Treehopper.Libraries.Displays
{
    using System;
    using System.Collections;
    using System.Threading.Tasks;
    using Utilities;

    /// <summary>
    /// Library for the HT16K33 16x8 LED driver, commonly used on Adafruit matrix displays
    /// </summary>
    public class Ht16k33 : LedDriver
    {
        private SMBusDevice dev;
        private BitArray data = new BitArray(128);
        private Package package;

        /// <summary>
        /// Construct a new HT16K33 16x8 LED driver
        /// </summary>
        /// <param name="i2c">The I2c port to use with this display.</param>
        /// <param name="address">The 7-bit I2c address to use</param>
        /// <param name="package">Which package is used</param>
        public Ht16k33(I2c i2c, byte address, Package package) : base((int)package, true, false)
        {
            dev = new SMBusDevice(address, i2c);
            dev.WriteByte(0x21);
            Brightness = 1.0;
            this.package = package;
        }

        /// <summary>
        /// Construct a new HT16K33 16x8 LED driver
        /// </summary>
        /// <param name="i2c">The I2c port to use with this display.</param>
        /// <param name="package">Which package to use</param>
        /// <param name="a0">The state of the A0 address pin</param>
        /// <param name="a1">The state of the A1 address pin</param>
        /// <param name="a2">The state of the A2 address pin</param>
        public Ht16k33(I2c i2c, Package package, bool a0 = false, bool a1 = false, bool a2 = false) : this(i2c, (byte)(0x70 | (a0 ? 1 : 0) | (a1 ? 1 : 0) << 1 | (a2 ? 1 : 0) << 2), package)
        {
        }

        /// <summary>
        /// The package types of the HT16K33. Larger packages support more LEDs.
        /// </summary>
        public enum Package
        {
            /// <summary>
            /// The SOP-20 package, supporting 64 (8x8) LEDs 
            /// </summary>
            Sop20_64Led = 64,

            /// <summary>
            /// The SOP-24 package, supporting 96 (12x8) LEDs
            /// </summary>
            Sop24_96Led = 96,

            /// <summary>
            /// The SOP-28 package, supporting 128 (16x8) LEDs
            /// </summary>
            Sop28_128Led = 128
        }

        private enum Commands
        {
            Brightness = 0xe0,
            Blink = 0x80
        }

        /// <summary>
        /// Flush the LED states to the HT16K33 driver
        /// </summary>
        /// <param name="force">Whether to force an update, even if no data appears to have changed</param>
        /// <returns></returns>
        public override Task Flush(bool force = false)
        {
            return dev.WriteBufferData(0x00, data.GetBytes());
        }

        internal override void LedBrightnessChanged(Led led)
        {
        }

        internal override void LedStateChanged(Led led)
        {
            int index = (16 * (led.Channel / ((int)package / 8))) + (led.Channel % ((int)package / 8));

            data.Set(index, led.State);
            if (AutoFlush)
            {
                // if we're autoflushing, only send the LED we need to update
                byte address = (byte)(index / 8);
                dev.WriteByteData(address, data.GetBytes()[address]).Wait();
            }
        }

        internal override void SetGlobalBrightness(double brightness)
        {
            if (brightness > 0)
            {
                dev.WriteByte((byte)((byte)Commands.Brightness | ((int)Math.Round(brightness * 16) - 1))).Wait();
                dev.WriteByte((byte)Commands.Blink | 0x01).Wait();
            }
            else
            {
                dev.WriteByte((byte)Commands.Blink | 0x00).Wait();
            }
        }
    }
}

File: Python/treehopper/libraries/displays/Ht16k33.py
from treehopper.api import I2C
from treehopper.utils.utils import bit_list_to_bytes
from treehopper.libraries import SMBusDevice
from treehopper.libraries.displays.led import LedDriver, Led


class Ht16k33(LedDriver):
    """Holtek HT16K33 LED driver

    """
    _command_brightness = 0xe0
    _command_blink = 0x80

    def __init__(self, i2c: I2C, package: int, a0=False, a1=False, a2=False, address=None):
        if not address:
            address = 0x70
            if a0:
                address |= 0x01
            if a1:
                address |= 0x02
            if a2:
                address |= 0x04

        LedDriver.__init__(self, num_leds=package, has_global_brightness_control=True,
                           has_individual_brightness_control=False)
        self._dev = SMBusDevice(address=address, i2c_module=i2c)
        self._dev.write_byte(0x21)
        self._package = package
        self._data = [False] * 128
        self._set_global_brightness(1.0)

    def _set_global_brightness(self, value: float):
        if self.brightness > 0.0:
            self._dev.write_byte(Ht16k33._command_brightness | round(self.brightness * 16) - 1)
            self._dev.write_byte(Ht16k33._command_blink | 0x01)
        else:
            self._dev.write_byte(Ht16k33._command_blink | 0x00)

    def flush(self, force=False):
        pass

    def _led_state_changed(self, led: Led):
        index = int(16 * int(led.channel / (self._package/8)) + (led.channel % (self._package/8)))
        self._data[index] = led.state
        if self.auto_flush:
            address = int(index / 8)
            self._dev.write_byte_data(address, bit_list_to_bytes(self._data)[address])

    def _led_brightness_changed(self, led: Led):
        pass


    class Package:
        sop20 = 64
        sop24 = 96
        sop28 = 128
